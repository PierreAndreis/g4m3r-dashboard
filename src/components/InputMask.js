import React from "react";

import Input from "./Input";

export const Masks = {
  secondsToMs: {
    in: value => {
      if (isNaN(Number(value))) return 0;
      // seconds to ms
      return value / 1000;
    },
    out: value => {
      if (isNaN(Number(value))) return 0;
      // ms to seconds
      return value * 1000;
    },
  },
};

class InputMask extends React.Component {
  onChange = e => {
    const value = e.target.value;
    this.props.onChange(this.props.mask.out(value));
  };

  render() {
    const { mask, value, ...props } = this.props;

    return (
      <Input
        {...props}
        // The order here matters. value and onChange can't be overwritten
        value={this.props.mask.in(this.props.value)}
        onChange={this.onChange}
      />
    );
  }
}

export default InputMask;
