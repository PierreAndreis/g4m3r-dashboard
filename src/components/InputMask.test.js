import React from "react";

import InputMask, { Masks } from "./InputMask";

import { render, fireEvent } from "react-testing-library";

const exampleMask = {
  in: value => Number(value) + 1,
  out: value => Number(value) - 1,
};

describe("InputMask", () => {
  it("Mounts", () => {
    const { container } = render(
      <InputMask mask={exampleMask} value="2" onChange={() => {}} />
    );

    expect(container).toMatchSnapshot();
  });

  it("masks the value", () => {
    const onChange = jest.fn();

    const { getByLabelText } = render(
      <InputMask
        mask={exampleMask}
        value={10}
        onChange={onChange}
        aria-label="input-test"
      />
    );

    const input = getByLabelText("input-test");

    // @fix: why string here? because "value" on the dom is always a string
    expect(input.value).toBe(String(exampleMask.in("10")));

    const event = {
      target: { value: "1" },
    };

    fireEvent.change(input, event);

    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith(exampleMask.out(event.target.value));
  });

  it("masks the placeholder", () => {
    const { getByLabelText } = render(
      <InputMask mask={exampleMask} placeholder={10} aria-label="input-test" />
    );

    const input = getByLabelText("input-test");

    // @fix: why string here? because "value" on the dom is always a string
    expect(input.placeholder).toBe(String(exampleMask.in("10")));
  });

  it("works without placeholder", () => {
    const { getByLabelText } = render(
      <InputMask mask={exampleMask} aria-label="input-test" />
    );

    const input = getByLabelText("input-test");

    expect(input.placeholder).toBe("");
  });
});

describe("Input Masks Helpers", () => {
  it("secondsToMs works properly", () => {
    expect(Masks.secondsToMs.in(10)).toMatchSnapshot();
    expect(Masks.secondsToMs.out(10)).toMatchSnapshot();
  });
});
