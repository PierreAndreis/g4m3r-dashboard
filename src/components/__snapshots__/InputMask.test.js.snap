// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Masks Helpers secondsToMs works properly 1`] = `0.01`;

exports[`Input Masks Helpers secondsToMs works properly 2`] = `10000`;

exports[`InputMask Mounts 1`] = `
.emotion-1 {
  width: 100%;
  box-sizing: border-box;
  height: 40px;
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  width: 100%;
  height: 100%;
  background: #eaeaea;
  border-radius: 5px;
  outline: 0;
  border: 2px solid transparent;
  padding: 2px 10px;
  box-sizing: border-box;
}

.emotion-0:focus {
  background: white;
  border-color: #7aaeff;
  box-shadow: 0 0 0 0.2rem rgba(122,174,255,0.15);
}

<div>
  <div
    class="emotion-1"
  >
    <div
      style="position: relative; width: 100%;"
    >
      <input
        class="emotion-0"
        value="3"
      />
    </div>
  </div>
</div>
`;
